### PR TITLE
[Merged by Bors] - Support AsBindGroup for 2d materials as well

### PR DESCRIPTION
**Pull request by johanhelsing**
Port changes made to Material in #5053 to Material2d as well.

This is more or less an exact copy of the implementation in bevy_pbr; I
simply pretended the API existed, then copied stuff over until it
started building and the shapes example was working again.

# Objective

The changes in #5053 makes it possible to add custom materials with a lot less boiler plate. However, the implementation isn't shared with Material 2d as it's a kind of fork of the bevy_pbr version. It should be possible to use AsBindGroup on the 2d version as well.

## Solution

This makes the same kind of changes in Material2d in bevy_sprite.

This makes the following work:

```rust
//! Draws a circular purple bevy in the middle of the screen using a custom shader

use bevy::{
    prelude::*,
    reflect::TypeUuid,
    render::render_resource::{AsBindGroup, ShaderRef},
    sprite::{Material2d, Material2dPlugin, MaterialMesh2dBundle},
};

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .add_plugin(Material2dPlugin::<CustomMaterial>::default())
        .add_startup_system(setup)
        .run();
}

/// set up a simple 2D scene
fn setup(
    mut commands: Commands,
    mut meshes: ResMut<Assets<Mesh>>,
    mut materials: ResMut<Assets<CustomMaterial>>,
    asset_server: Res<AssetServer>,
) {
    commands.spawn_bundle(MaterialMesh2dBundle {
        mesh: meshes.add(shape::Circle::new(50.).into()).into(),
        material: materials.add(CustomMaterial {
            color: Color::PURPLE,
            color_texture: Some(asset_server.load("branding/icon.png")),
        }),
        transform: Transform::from_translation(Vec3::new(-100., 0., 0.)),
        ..default()
    });

    commands.spawn_bundle(Camera2dBundle::default());
}

/// The Material2d trait is very configurable, but comes with sensible defaults for all methods.
/// You only need to implement functions for features that need non-default behavior. See the Material api docs for details!
impl Material2d for CustomMaterial {
    fn fragment_shader() -> ShaderRef {
        "shaders/custom_material.wgsl".into()
    }
}

// This is the struct that will be passed to your shader
#[derive(AsBindGroup, TypeUuid, Debug, Clone)]
#[uuid = "f690fdae-d598-45ab-8225-97e2a3f056e0"]
pub struct CustomMaterial {
    #[uniform(0)]
    color: Color,
    #[texture(1)]
    #[sampler(2)]
    color_texture: Option<Handle<Image>>,
}
```

### CODE REVIEW

File: crates/bevy_sprite/src/mesh2d/material.rs [hunk 1]
```diff
@@ -39,36 +45,85 @@ use crate::{
...
+/// var<uniform> material: CustomMaterial;
+/// [[group(1), binding(1)]]
+/// var color_texture: texture_2d<f32>;
+/// [[group(1), binding(2)]]
+/// var color_sampler: sampler;
+/// ```
+pub trait Material2d: AsBindGroup + Send + Sync + Clone + TypeUuid + Sized + 'static {
     /// Returns this material's vertex shader. If [`None`] is returned, the default mesh vertex shader will be used.
     /// Defaults to [`None`].
     #[allow(unused_variables)]
```
Comment (devil-ira) [inline crates/bevy_sprite/src/mesh2d/material.rs:117]: These allow annotations can be removed :)

File: crates/bevy_sprite/src/mesh2d/material.rs [hunk 2]
```diff
@@ -366,9 +374,151 @@ pub fn queue_material2d_meshes<M: SpecializedMaterial2d>(
     }
 }
 
-/// A component bundle for entities with a [`Mesh2dHandle`] and a [`SpecializedMaterial2d`].
+/// Data prepared for a [`Material`] instance.
```
Comment (IceSentry) [inline crates/bevy_sprite/src/mesh2d/material.rs:377]: This is the fix for the failing doc test
```suggestion
/// Data prepared for a [`Material2d`] instance.
```

Comment (johanhelsing) [reply crates/bevy_sprite/src/mesh2d/material.rs]: Thanks

File: crates/bevy_sprite/src/mesh2d/material.rs (hunk 1 above)
Comment (johanhelsing) [reply crates/bevy_sprite/src/mesh2d/material.rs]: Yay :) Done.

**Review by IceSentry — APPROVED** [review]

**Review by cart — APPROVED** [review]
Great work. No comments!

**Comment by cart** [comment]
bors retry

**Comment by cart** [comment]
(failure was due to a nightly incremental issue. retrying might help)

**Comment by mockersf** [comment]
bors retry

**Pull request closed without merging**